livekit/_ffi_client.py

```diff
@@ -68,26 +69,50 @@ def ffi_event_callback(data_ptr: ctypes.POINTER(ctypes.c_uint8),  # type: ignore
     event_data = bytes(data_ptr[:int(data_len)])
     event = proto_ffi.FfiEvent()
     event.ParseFromString(event_data)
+    ffi_client._event_received(event)
 
-    ffi_client._event_loop.call_soon_threadsafe(dispatch_event, event)
 
-
-def dispatch_event(event: proto_ffi.FfiEvent) -> None:
-    which = str(event.WhichOneof('message'))
-    ffi_client.emit(which, getattr(event, which))
-
-
-class FfiClient(EventEmitter):
+class FfiClient:
     def __init__(self) -> None:
         super().__init__()
-        self._event_loop = asyncio.get_event_loop()
 
+        self.subscribers: List[asyncio.Queue[proto_ffi.FfiEvent]] = []
+        self.lock = threading.RLock()
+
+        # initialize request
         req = proto_ffi.FfiRequest()
         cb_callback = int(ctypes.cast(
             ffi_event_callback, ctypes.c_void_p).value)  # type: ignore
         req.initialize.event_callback_ptr = cb_callback
         self.request(req)
 
+    async def wait_for_event(self, fnc: Callable[[proto_ffi.FfiEvent], bool]) \
+            -> proto_ffi.FfiEvent:
+        queue = self.subscribe()
+        try:
+            while True:
+                event = await queue.get()
+                if fnc(event):
+                    return event
```

Comment:
@theomonnom when do we remove the event from the queue?

Reply:
get is removing it

Reply:
Isn't this a problem then where we remove it even if fn returns False?